### PR TITLE
Add Zendesk chat module

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
Add Zendesk pre-sales chat widget to `my-jetpack`

## Proposed changes:

This adds a conditionally rendered (based on date and time) Zendesk chat widget by embedding the chat widget script

It is configured to only be visible during chat available hours

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

P2: pbNhbs-5cD-p2

## Does this pull request change what data or activity we track or use?

Depends on how that is interpreted, I'm sure the chats are all saved somewhere, so in that sense, yes. No other trackers are being added in addition to that though

## Testing instructions:

1. Checkout this branch
2. Head to `/wp-admin/admin.php?page=my-jetpack`
3. If you are looking at this on a Monday-Friday between 9:00 UTC and 23:00 UTC, ensure the widget shows up in the bottom right corner
 
image

5. You can fudge the numbers in the `utils.ts` script to test when outside of hours if you'd like, but the new automated tests specifically test this, so you could just check if the way the tests are written look good

### CODE REVIEW

File: projects/js-packages/components/components/zendesk-chat/README.md
Comment (jboland88) [inline projects/js-packages/components/components/zendesk-chat/README.md]: > to any page the componet is added typeof

I'm not sure I understand this.

File: projects/js-packages/components/components/zendesk-chat/index.tsx
Comment (jboland88) [inline projects/js-packages/components/components/zendesk-chat/index.tsx:11]: I suppose by memoizing this, it will just check if it should show at the first render, but wouldn't re-check on subsequent renders (based on the current time). 🤔 That seems fine, but just wanted to note it.

File: projects/js-packages/components/components/zendesk-chat/utils.ts
Comment (jboland88) [inline projects/js-packages/components/components/zendesk-chat/utils.ts]: Nitpick - I wonder if `array.includes()` would be easier to parse here?

File: projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx
Comment (jboland88) [inline projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx:15]: How does the zendesk script work? Does it expect that the chat container exists before the script is added/ run? Or could the chat container be added later on? Is the container here just an anchor for the script to be added into?

Something about assembling the `<script>` element in React feels like an anti-pattern. Would there be a workable way to load the script via `wp_enqueue_script`? I don't think this is an invalid approach - but curious if we can load the script from the backend? Maybe not?

**Review by jboland88 — CHANGES_REQUESTED** [review]
This tests well, but I wonder if doing this in React may be making some things more complicated?

If all we need to do is load the script, I wonder if we could implement this in PHP with a conditional call to `wp_enqueue_script` for the My Jetpack page?

File: projects/js-packages/components/components/zendesk-chat/README.md
Comment (CodeyGuyDylan) [reply projects/js-packages/components/components/zendesk-chat/README.md]: Looks like a weird autofill is the culprit of this 😅 
image

File: projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx
Comment (CodeyGuyDylan) [reply projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx:15]: For some context, this implementation was to somewhat emulate the [implementation on Calypso](https://github.com/Automattic/wp-calypso/tree/trunk/client/components/presales-zendesk-chat), and while I think there is some value in having similar implementations across the two for ease of updating, I agree that if we can get it to work on the backend with a simple `wp_enque_script`, it may be a better way to go about it

I did not think of doing it this way, so I have not confirmed that it will/will not work. In Claypso they use a function called `loadScript` which doesn't really exist, that I know of or could find, in this codebase, and it is a semi-complicated function that I didn't just want to throw into this PR. 

I'll go ahead and see if loading it from the backend can meet our needs if you still think that's the better approach after laying out the context

Comment (CodeyGuyDylan) [reply projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx:15]: > Is the container here just an anchor for the script to be added into?

It is my understanding that the container is just an anchor for the script to be added into

File: projects/js-packages/components/components/zendesk-chat/utils.ts
Comment (CodeyGuyDylan) [reply projects/js-packages/components/components/zendesk-chat/utils.ts]: I agree, I have updated this in case we don't go the backend route

**Comment by CodeyGuyDylan** [comment]
@jboland88 and I threw together a backend version of this (just loading the script, no tests or conditionals), and we got it to work with the following code in the backend of `my-jetpack`

```php
// New method in class-initializer.php
public static function add_id_to_zendesk_widget_script( $tag, $handle, $src ) {
   if ( 'presales-zendesk-chat' === $handle ) {
	$tag = '<script type="text/javascript" src="' . esc_url( $src ) . '" id="ze-snippet"></script>';
   }

   return $tag;
}

// In enqueue_scripts in class-initializer.php
$zendesk_script_args = array(
	'key' => urlencode( 'd729d42c-b547-4750-a6f6-8b30534a5f12' ),
);
wp_enqueue_script( 'presales-zendesk-chat',  add_query_arg( $zendesk_script_args, 'https://static.zdassets.com/ekr/snippet.js', array( 'my_jetpack_main_app' ) ) );

add_filter( 'script_loader_tag', array( __CLASS__, 'add_id_to_zendesk_widget_script' ), 10, 3 );
```

We decided against using it for now for a couple of reasons
1. Refactoring this correctly would involve creating a new class so it's not lumped in with the main backend class, as well as adding conditionals and writing tests again. This seemed like a lot of work for something we already have a useable solution for
2. This method would be less reusable than the React version. With the current implementation, we could import and drop that component anywhere and it would just work. With the backend version, we'd have to repeat the code on the other package/plugin to get it to load there
3. We had to inject an ID with a filter in order for Zendesk to load the widget, not ideal

Reasons we thought about doing it on the plugin backend
1. Loading a script in React seems not very 'react-y'
2. Loading the script on the backend seems like better practices overall

File: projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx
Comment (IanRamosC) [reply projects/js-packages/components/components/zendesk-chat/zendesk-chat-script.tsx:15]: I see how this might seem like an anti-pattern - and I'd prefer to have it loaded in the backend as well - but this is actually a common way to embed external scripts into React components. Considering what y'all have discussed, I also think we should go with this solution.

File: projects/js-packages/components/components/zendesk-chat/index.tsx
Comment (IanRamosC) [inline projects/js-packages/components/components/zendesk-chat/index.tsx:10]: I'd love if we could fetch the available chat times from the backend in the future.

**Review by IanRamosC — DISMISSED** [review]
This tests well and seems consistent with the implementation in Calypso.

File: projects/js-packages/components/components/zendesk-chat/index.tsx
Comment (CodeyGuyDylan) [reply projects/js-packages/components/components/zendesk-chat/index.tsx:10]: This is a good idea, I believe an endpoint is being worked on to that provides this info, we can make that an addition later

**Review by jeherve — DISMISSED** [review]
This tests well for me. I'm not sure how I feel about hardcoding opening days and times in the plugin, so folks with outdated versions of the package may see an open form when it's supposed to be closed at some point in the future, but maybe that's something we can iterate upon in the future?

**Comment by CodeyGuyDylan** [comment]
> This tests well for me. I'm not sure how I feel about hardcoding opening days and times in the plugin, so folks with outdated versions of the package may see an open form when it's supposed to be closed at some point in the future, but maybe that's something we can iterate upon in the future?

I agree with this, I believe an endpoint with that information is in the works: D99766-code, so I can add this task to our board and if this gets merged, depending on priorities at the time, I can try and get this updated before the next code freeze

**Comment by CodeyGuyDylan** [comment]
@jboland88 Just re-requesting review because I believe it is blocking merge

File: projects/js-packages/components/components/zendesk-chat/index.tsx
Comment (jboland88) [reply projects/js-packages/components/components/zendesk-chat/index.tsx:10]: It looks like that endpoint might have been completed in D99766-code - I see you have already mentioned creating a follow-up task to revisit this :)

**Review by jboland88 — DISMISSED** [review]
LGTM! Noticed that the changes in D99766-code are completed - but I see you're already following up on that in a separate issue.

**Comment by CodeyGuyDylan** [comment]
> LGTM! Noticed that the changes in D99766-code are completed - but I see you're already following up on that in a separate issue.

Yep, it's here: 1198218726984184-as-1204094891607626/f 😄

**Review by jboland88 — APPROVED** [review]
Re-approving after trunk merge - only versioning was updated :)

**Pull request merged**
